src/infrasys/pint_quantities.py

```diff
@@ -0,0 +1,206 @@
+            - An unknown unit is provided.
+            - An invalid type is provided for the value.
+        TypeError
+            If an error occurs in the unit registry or context, which is not caused by user input.
+            This exception is not propagated as a `pydantic.ValidationError`.
+        """
+        # NOTE: `self.ureg` when passed returns the rigth type
+        if not isinstance(input_value, Quantity):
+            input_value = self.ureg(input_value)
+
+        if isinstance(input_value, Number | list):
+            input_value = input_value * self.units
+        try:
+            input_value = input_value.to(self.units)
+        except pint.DimensionalityError:
+            msg = f"Dimension mismatch from {input_value.units} to {self.units}"
+            raise ValueError(msg)
+        return input_value
+
+    def serialize(
+        self,
+        value: Quantity,
+        info: core_schema.SerializationInfo | None = None,
+        *,
+        to_json: bool = False,
+    ) -> dict | str | Quantity:
+        """
+        Serialize a `PydanticPintQuantity`.
+
+        Parameters
+        ----------
+        value : pint.Quantity
+            The quantity to serialize. This should be a `pint.Quantity` object.
+        info : core_schema.SerializationInfo, optional
+            The serialization information provided by the Pydantic schema. Default is `None`.
+        to_json : bool, optional
+            If `True`, serializes to a JSON-compatible format (string or dictionary).
+            If `False` (default), returns the quantity as is (i.e., `pint.Quantity`).
+
+        Returns
+        -------
+        dict, str, or pint.Quantity
+            The serialized representation of the quantity.
+            - If `to_json` is `True`, the quantity is serialized to a JSON-compatible string or dictionary.
+            - If `to_json` is `False`, returns the original `pint.Quantity`.
+            - If `ser_mode=='dict'` or `info.mode=='dict'` a dictionary with magniude and units.
+
+        Notes
+        -----
+        This method is useful when working with `PydanticPintQuantity` fields outside
+        of Pydantic models, as it allows control over the serialization format
+        (e.g., JSON-compatible representation).
+        """
+        # If we pass the to_json flag, we serialize it as json, but sometimes the pydantic info can have the to_json
+        # enabled. In that case so we add a check of that as well
+        info_to_json = False
+        if info is not None and info.mode_is_json():
+            info_to_json = True
+        to_json = to_json or info_to_json
+
+        if self.ser_mode == "dict" or (info is not None and info.mode == "dict"):
+            return {
+                "magnitude": value.magnitude,
+                "units": value.units if not to_json else f"{value.units}",
+            }
+
+        if self.ser_mode == "str" or to_json:
+            return f"{value}"
+
+        return value
```

Comment:
I'm not following this case. It's is basically bypassing serialization. When would we do this?

Reply:
Not very useful in most of the cases. I think only if you want to serialize the unit like "10 volt"

Reply:
That's what happens on line 158, right? (the instance of Quantity is converted to a string). In this case you are returning the instance of Quantity, making the method a no-op. Am I missing something?